sklearn/calibration.py

```diff
@@ -658,3 +660,232 @@ def calibration_curve(y_true, y_prob, *, normalize=False, n_bins=5,
+    method exists, :term:`decision_function` will be used and the output
+    confidence scores will be min-max scaled to the range [0,1].
+
+    Extra keyword arguments will be passed to :func:`matplotlib.pyplot.plot`.
+
+    Read more in the :ref:`User Guide <calibration>`.
+
+    Parameters
+    ----------
+    estimator : estimator instance
+        Fitted classifier or a fitted :class:`~sklearn.pipeline.Pipeline` in
+        which the last estimator is a classifier.
+
+    X : {array-like, sparse matrix} of shape (n_samples, n_features)
+        Input values.
+
+    y : array-like of shape (n_samples,)
+        Binary target values.
+
+    n_bins : int, default=5
+        Number of bins to discretize the [0, 1] interval into when calculating
+        the calibration curve. A bigger number requires more data.
+
+    strategy : {'uniform', 'quantile'}, default='uniform'
+        Strategy used to define the widths of the bins.
+
+        - `'uniform'`: The bins have identical widths.
+        - `'quantile'`: The bins have the same number of samples and depend on
+          predicted probabilities.
+
+    name : str, default=None
+        Name for labeling curve. If `None`, the name of the estimator is used.
+
+    brier_score : bool, default=True
+        If `True`, include Brier score in legend.
+
+    ref_line : bool, default=True
+        If `True`, plots a reference line representing a perfectly calibrated
+        classifier.
+
+    ax : matplotlib axes, default=None
+        Axes object to plot on. If `None`, a new figure and axes is created.
+
+    **kwargs : dict
+        Keyword arguments to be passed to :func:`matplotlib.pyplot.plot`.
+
+    Returns
+    -------
+    display : :class:`~sklearn.calibration.CalibrationDisplay`.
+        Object that stores computed values.
+
+    Examples
+    --------
+    >>> import matplotlib.pyplot as plt                     # doctest: +SKIP
+    >>> from sklearn import (datasets, calibration, model_selection,
+    ...                      linear_model)
+    >>> X, y = datasets.make_classification(random_state=0)
+    >>> X_train, X_test, y_train, y_test = model_selection.train_test_split(
+    ...     X, y, random_state=0)
+    >>> clf = linear_model.LogisticRegression(random_state=0)
+    >>> clf.fit(X_train, y_train)
+    LogisticRegression(random_state=0)
+    >>> metrics.plot_calibration_curve(clf, X_test, y_test)  # doctest: +SKIP
+    >>> plt.show()                                           # doctest: +SKIP
+    """
+    check_matplotlib_support("plot_calibration_curve")
+    binary_error = "Only binary classification is supported."
+
+    if not is_classifier(estimator):
+        raise ValueError("The estimator parameter should be a fitted binary "
+                         "classifier")
+
+    predict_proba = getattr(estimator, 'predict_proba', None)
+    decision_function = getattr(estimator, 'decision_function', None)
+    prediction_method = predict_proba or decision_function
```

Comment:
in `_CalibratedClassifier`, decision_function has priority over predict_proba

Reply:
Though I'm a bit confused now: why would one want to pass the result of the decision_function into `calibration_curve`?

A classifier may be calibrated based on its `predict_proba` or `decision_function`, but ultimately, what gets calibrated are the probabilities, not the decision function. So I'm not sure that what is done below (converting a decision function into a proba) makes sense

Reply:
Thanks for raising this. I'm not too familiar with this but I think SVMs don't produce probabilities. However these can be calculated assuming the distance to hyperplane values are monotonically related to a confidence level. E.g, via Platt scaling, which uses a logistic equation. From [Platts paper](https://www.researchgate.net/profile/John_Platt/publication/2594015_Probabilistic_Outputs_for_Support_Vector_Machines_and_Comparisons_to_Regularized_Likelihood_Methods/links/004635154cff5262d6000000.pdf):

> This sigmoid model is equivalent to assuming that the output of the SVM is proportional to the log-likelihood of a positive training example.

I think using `_CalibratedClassifier` on a SVM (or setting `probability=True` in the SVM estimator) transforms the `decision_function` outputs into probabilities. I don't know if `_CalibratedClassifier` *should* be the only way for users to obtain probabilities or whether scaling the `decision_function` should also be an option?

I initially implemented only allowing `predict_proba` but I thought that since `_CalibratedClassifier` is computationally expensive, as it involves cross-validation, I would allow `decision_function` and do simple scaling. I really don't know though. Note the [calibration comparison](https://github.com/scikit-learn/scikit-learn/blob/master/examples/calibration/plot_compare_calibration.py#L100-L102) example does min-max scaling on the decision scores from `LinearSVC` (which interestingly, is one of the few SVMs without a `probability` parameter).

Also for the estimators with both a `decision_function` and `predict_proba` method, I thought it would be better to 'always' use the `predict_proba`, instead of transforming the `decision_function`.

Finally, apart from SVMs are there other estimators that only have a `decision_function` method...?

Reply:
I'll try to answer / comment each point more in details below. Things are a bit clearer in my mind after I had my morning coffee so I can sum up my POV with this:

`calibration_curve` expects predicted probabilities so we should only allow estimators that can output probabilities here. We should not try to get probabilities by scaling the `decision_function`.
(Note that calibration can perfectly be done based on the decision function, but that's another matter.)

----

> I think using _CalibratedClassifier on a SVM (or setting probability=True in the SVM estimator) transforms the decision_function outputs into probabilities

You're correct. The way it's done when using `probability=True` is slightly different from using our `CalibratedClassifierCV` (details here https://github.com/scikit-learn/scikit-learn/pull/16167), but that's not important here

> don't know if _CalibratedClassifier should be the only way for users to obtain probabilities or whether scaling the decision_function should also be an option

I feel like getting probabilities from scaling the decision function (as done in the example you mentioned) is somewhat of a hack. The proper way to get probability from a decision function should be calibration.

> for the estimators with both a decision_function and predict_proba method, I thought it would be better to 'always' use the predict_proba, instead of transforming the decision_function

I definitely agree (I was just confused in https://github.com/scikit-learn/scikit-learn/pull/17443#discussion_r439735610). As mentioned above I even think we should not even consider the decision function here.

> Finally, apart from SVMs are there other estimators that only have a decision_function method...?

Depending on the loss, some instances of the `SGDClassifier` (as e.g. the Perceptron, or using any regression loss) don't have an interpretable probabilitstic output (see https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html#sklearn.linear_model.SGDClassifier.predict_proba). In order to get a `predict_proba` method, these classifiers should be calibrated first (using the ouput of the decision function).

Reply:
Thanks for your thoughtful reply! I will revert to allow `predict_proba` only.

Reply:
I will also change the examples to use `CalibratedClassifierCV` with `LinearSVC` - is that okay?

Reply:
the issue is that the examples with LInearSVC are meant to illustrate how poorly it is calibrated by default (so calibrating it would prevent us from illustrating this)

I think that both in https://scikit-learn.org/stable/auto_examples/calibration/plot_compare_calibration.html#sphx-glr-auto-examples-calibration-plot-compare-calibration-py and in  https://scikit-learn.org/stable/auto_examples/calibration/plot_calibration_curve.html#sphx-glr-auto-examples-calibration-plot-calibration-curve-py we should clarify that for un-calibrated LinearSVC, we pretend to have probabilities by scaling the decision function. (The UG would also need an update). I guess that also means you'll need to use the Display object with LinearSVC instead of relying on `plot_calibration_curve`

We may also consider just removing it from https://scikit-learn.org/stable/auto_examples/calibration/plot_compare_calibration.html#sphx-glr-auto-examples-calibration-plot-compare-calibration-py 

Maybe @ogrisel can weight in.

Reply:
Yes good point  - I remember now that I had used `CalibrationDisplay` in the example when I implemented allowing `predict_proba` only (I need my coffee).

I'll wait and see if @ogrisel replies before I do anything. Thanks!

Reply:
I really like the educational value of showing how naive min-max scaling of decision function leads to poorly calibrated SVMs in the example.

I am in favor in keeping this feature but maybe warn with a message such as :

```python
f"Estimator {estimator_name} has no predict_proba method. Naively mapping the output of "
f" decision_function to the [0, 1] range using a min-max scaling strategy instead."
```

And maybe we could add an option named `naive_predict_proba="warn"` with possible values in `{'warn', 'min-max', 'logistic_sigmoid'}` to address @thomasjpfan 's comment: https://github.com/scikit-learn/scikit-learn/pull/17443/files#r439100411.

Reply:
I really like the educational value of showing how naive min-max scaling of decision function leads to poorly calibrated SVMs in the example.

I am in favor in keeping this feature but maybe warn with a message such as :

```python
f"Estimator {estimator_name} has no predict_proba method. Naively mapping the output of "
f" decision_function to the [0, 1] range using a min-max scaling strategy instead."
```

And maybe we could add an option named `naive_predict_proba="warn"` with possible values in `{'warn', 'min-max', 'logistic_sigmoid'}` to address @thomasjpfan 's comment: https://github.com/scikit-learn/scikit-learn/pull/17443/files#r439100411.

Reply:
I agree about the pedagogical value but I don't think the mapping should be part of the plotting function, which should not contain any calibration-related logic.

If anything, the mapping could be part of the calibrator estimator `CalibratedClassifierCV`.

Though I'd rather just let users manually do it: there's an infinite number of ways to map arbitrary values into [0, 1], and since none of them is recommended in this context, I feel like we shouldn't have built-in support for an of them.

Reply:
Point taken.

Maybe we could just do in the example it-self:

```python
class NaivelyCalibratedLinearSVC(LinearSVC):

    def predict_proba(self, X):
        unscaled_predictions = self.decision_function(X)
        # TODO: put the min-max scaling here
```

and remove it from the utility function. They would raise a `TypeError` or `AttributeError` if the provided classifier does not implement `predict_proba` it-self.

Reply:
I'd be happy with that!

Reply:
@lucyleeow it seems we are reaching a consensus. What's your take on the above suggestion?

Reply:
@ogrisel for the example, do you prefer defining the new estimator class with a `predic_proba` over using the display object `CalibrationDisplay` and manually calculating and setting the attributes of the class instance?

Reply:
e.g., the latter:
```
        y_prob = clf.decision_function(X_test)
        y_prob = (y_prob - y_prob.min()) / (y_prob.max() - y_prob.min())
        prob_true, prob_pred = calibration_curve(y, y_prob, n_bins=10)
        brier = brier_score_loss(prob_true, prob_pred)

        viz = CalibrationDisplay(
            prob_true, prob_pred, y_prob, estimator_name=name
        )
        viz.plot(ax=ax1)
```

Reply:
I would rather use the `plot_calibration_curve` API in the example. Subclassing the `LinearSVC`  is fine to me.